### ISSUE TITLE
Jinja format for bumping version of <distance> steps

### ISSUE THREAD

**Issue by kalfa** [mtkennerly/poetry-dynamic-versioning#11]
Hi,
my workflow bumps base version by N patch-versions, where N is the distance from the last tagged version.

I wrote this macro which does what I need, and wanted to share with you and the users, if this helps. 
In my case it works out of the box, but discards any other elements which is not base or distance, which can clearly be a limitation for other people

do you think we could have instead a function which does that (in python)?

If there is a official way to extend your 'plugin', it would be great, so I can add this function just to my extention and keep upstream simple.

```
[tool.poetry-dynamic-versioning]
  enable = true
  style = "pep440"

  format-jinja = """
{%- macro bump_n_versions(base, distance) -%}
  {%- if distance|int >= 2 -%}
    {%- set bumped_base = bump_version(base) -%}
    {{- bump_n_versions(bumped_base, distance-1) -}}
  {%- elif distance|int == 1 -%}
    {%- set bumped_base = bump_version(base) -%}
    {{- bumped_base -}}
  {%- else -%} {# distance==0 #}
    {{- base -}}
  {%- endif -%}
{%- endmacro -%}

{{- bump_n_versions(base, distance) -}}
"""
```

**Comment by mtkennerly** [comment]
Thanks for sharing! It's helpful to get a sense of the different workflows that people use.

You should be able to simplify that code a bit:

```
[tool.poetry-dynamic-versioning]
enable = true
style = "pep440"
format-jinja = """
    {%- set ns = namespace(version=base) -%}
    {%- for i in range(distance) -%}
        {%- set ns.version = bump_version(ns.version) -%}
    {%- endfor -%}
    {{- ns.version -}}
"""
```

I hadn't really considered extensions/plugins to the plugin itself. Would you just want a way to define functions in a separate Python file in your repository, rather than defining them in the Jinja itself? I have mixed feelings about that; on one hand, I could see it being helpful for things like syntax highlighting, but on the other hand, I feel like `format-jinja` would generally be simple enough that it'd be overkill to split into another file. Could you tell me a bit more about what you had in mind?

**Comment by kalfa** [comment]
Oh, I know, the code was just a lazy approach to sort out a functionality. not a jinja-master, had to look at macros. thank for rewriting it.

I have two requirements, which I guess are more requirements for the larger system (pip/pep517 and poetry).

first, defining custom functions like the one in jinja. yes, it's possible to write things in jinja and definitely a good interim, but normally it's good/easier to write the logic in python and just call it. it's not presentation after all, it's business logic.
althogh, it's really nice to be able to write it in jinja, so I don't have to release a package with my code, add it to the building-reqs and then have your package to access it.

the second requirement, which is connected to the first and makes actually the first stronger, is to be able to manage multiple repositories with the same pyproject core functions.

this goes toward several solution, one is the one mentioned in the first one.

use case: I have a project composed by several python packages. it's a large project.
each package will be built with pyproject (now it uses setup.py).
in the old setup() world I wrote a packge on which the building depended, and had the logic to compute things. for example I used to have the implementaion of the logic your plugin+the jinja logic implemented.

now, since this is re-used by several setup() in sevearl repositories, I need to have a central place to have the logic. I don't want to re-define the logic in each repo, right?

I'm looking for the same functionality here. how can I have the logic of the format-jinja code in all of my pyproject.toml without having to re-define the same liens over and over (and amend them over and over if there is a change to apply)?

one approach is

```
[tool.poetry-dynamic-versioning]
enable = true
extra_functions = [
   'my.extra.function.module:bump_version_by_n',
   'my.extra.function.module:tell_me_a_joke_before_bumping_version,
]
format-jinja = "{{- bump_version_by_n(base) -}}"

[build-system]
  requires = [
         "my-extra-functions",
         "poetry-dynamic-versioning",
         "poetry",
          "wheel"
  ]
  build-backend = "poetry.masonry.api"
```

another approach is to  be able to wrap my plugin using yours.

```
[tool.my-project.dynamic-versioning]
enable = true

[build-system]
  requires = [
         "my-project-dynamic-versioning",
         "poetry",
          "wheel"
  ]
  build-backend = "poetry.masonry.api"
```
This could add more configurability (which might later be intergrated into your project), or hide details of the configuraton (which is my use case) and delegate it to the wrapping package instead, which is shared among multiple packages.

and this would use your project to extend it. I was looking at the way I could do it.
I am exploring the related peps right now from an advanced user POV, and see how can we really exploit them.

I guess to achieve this we would need a public API to use.

the main use cases are anyway:
- expand extendability for features
- hide common repeated configurations

**Comment by mtkennerly** [comment]
I think the `extra_functions` approach sounds good (I might call it `format-jinja-imports`).

For now, I'd be a little hesitant to make the plugin more broadly extensible until there's a clear use case that doesn't belong in the plugin itself. In terms of minimizing repeated configuration, I would need to think some more about how to handle that; my first thought is stacking settings from multiple pyproject.toml files, but I haven't really thought through that.

**Comment by mtkennerly** [comment]
This is now available in 0.8.0 as `format-jinja-imports`. Please see the README for more info.
